sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/CoroutinesTimeoutApiUnusedCheck.kt

```diff
@@ -0,0 +1,76 @@
+package org.sonarsource.kotlin.checks
+
+import org.jetbrains.kotlin.psi.KtCallExpression
+import org.jetbrains.kotlin.psi.KtDotQualifiedExpression
+import org.jetbrains.kotlin.psi.KtElement
+import org.jetbrains.kotlin.psi.KtExpression
+import org.jetbrains.kotlin.psi.KtProperty
+import org.jetbrains.kotlin.psi.psiUtil.siblings
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.bindingContextUtil.getReferenceTargets
+import org.jetbrains.kotlin.resolve.calls.model.ResolvedCall
+import org.sonar.check.Rule
+import org.sonarsource.kotlin.api.CallAbstractCheck
+import org.sonarsource.kotlin.api.FunMatcher
+import org.sonarsource.kotlin.api.KOTLINX_COROUTINES_PACKAGE
+import org.sonarsource.kotlin.api.SecondaryLocation
+import org.sonarsource.kotlin.api.matches
+import org.sonarsource.kotlin.converter.KotlinTextRanges.textRange
+import org.sonarsource.kotlin.plugin.KotlinFileContext
+
+private const val MESSAGE = """Use "withTimeoutOrNull { }" instead of manual delayed cancellation"""
+
+private val DELAY_MATCHER = FunMatcher(name = "delay", qualifier = KOTLINX_COROUTINES_PACKAGE)
+private val LAUNCH_ASYNC_MATCHER = FunMatcher(qualifier = KOTLINX_COROUTINES_PACKAGE) {
+    withNames("launch", "async")
+}
+
+@Rule(key = "S6316")
+class CoroutinesTimeoutApiUnusedCheck : CallAbstractCheck() {
+    override fun functionsToVisit() = listOf(FunMatcher(supertype = "$KOTLINX_COROUTINES_PACKAGE.Job", name = "cancel"))
+
+    override fun visitFunctionCall(
+        cancelCallExpr: KtCallExpression,
+        resolvedCancelCall: ResolvedCall<*>,
+        kotlinFileContext: KotlinFileContext
+    ) {
+        val bindingContext = kotlinFileContext.bindingContext
+
+        val cancelCallCalleeExpression = cancelCallExpr.calleeExpression ?: return
```

Comment:
Not obligatory, but according to my experience: It seems that if you've already matched some function signature, `caleeExpression` won't be null, so you can use non-null assertion safely.